src/operator/nn/mkldnn/mkldnn_rnn_impl.h

```diff
@@ -0,0 +1,739 @@
+          (*wh_memory)[layer_index], (*bias_memory)[layer_index],
+          (*y_memory)[layer_index],
+         (*hcy_memory)[layer_index], null_memory_);
+    rnn_forward_prim->push_back(rnn_prim);
+  }
+  MKLDNNStream::Get()->RegisterPrim((*rnn_forward_prim)[layer_index]);
+  MKLDNNStream::Get()->Submit();
+
+  if (state_outputs) {
+    DType* dst_hcy = reinterpret_cast<DType *> ((*hcy_memory)[layer_index].get_data_handle());
+    if (mode == rnn_enum::kLstm) {
+      for (int l = 0; l < L; l++) {
+        offset1 = l * single_cell_size;
+        offset2 = l * nstates * single_cell_size;
+        #pragma omp parallel for num_threads(omp_threads)
+        for (int n = 0; n < single_cell_size; n++) {
+          hy_ptr[offset1 + n] = dst_hcy[offset2 + n];
+          cy_ptr[offset1 + n] = dst_hcy[offset2 + n + single_cell_size];
+        }
+      }
+    } else {
+      #pragma omp parallel for num_threads(omp_threads)
+      for (int n = 0; n < L * single_cell_size; n++) {
+        hy_ptr[n] = dst_hcy[n];
+      }
+    }
+  }
+}
+
+template <typename DType>
+void MKLDNNRNNForward(bool state_outputs,
+                      const int L,
+                      const int D,
+                      const int T,
+                      const int N,
+                      const int I,
+                      const int H,
+                      DType* x_ptr,
+                      DType* hx_ptr,
+                      DType* cx_ptr,
+                      DType* w_ptr,
+                      DType* b_ptr,
+                      DType* y_ptr,
+                      DType* hy_ptr,
+                      DType* cy_ptr,
+                      std::vector<mkldnn::memory> *concat_weight_memory,
+                      std::vector<mkldnn::memory> *concat_iter_memory,
+                      std::vector<mkldnn::memory> *x_memory,
+                      std::vector<mkldnn::memory> *hcx_memory,
+                      std::vector<mkldnn::memory> *wx_memory,
+                      std::vector<mkldnn::memory> *wh_memory,
+                      std::vector<mkldnn::memory> *bias_memory,
+                      std::vector<mkldnn::memory> *y_memory,
+                      std::vector<mkldnn::memory> *hcy_memory,
+                      std::vector<primitive> *rnn_forward_prim,
+                      bool *has_cache,
+                      int dtype,
+                      bool is_train,
+                      int mode) {
+  int ngates = 0, nstates = 0;
+  GetMKLDNNRNNAlgo(mode, &ngates, &nstates);
+  const int b_size = 2 * H * ngates * D;
+  const int cell_size = N * H * D;
+  //  First layer
+  int w_size = (I + H) * H * ngates * D;
+  auto cpu_engine = CpuEngine::Get()->get_engine();
+  auto null_memory_ = null_memory(cpu_engine);
+  DType* tmpNull = NULL;
+  // when D = 1 and I == H, L layers can be fused together
+  if (D == 1 && I == H) {
+    MKLDNNRNNForwardUnidi(state_outputs, L, T, N, I, H, x_ptr, &null_memory_,
```

Comment:
Too many sub-function for different cases and I'm afraid it's really error-prone.